### PR TITLE
feat: Add Windows Service support

### PR DESCRIPTION
**Pull request by fkollmann**
This PR adds support for running the Cloud SQL Proxy as Windows service.

More details on how to install, can be found in the `windows-service-guide.md` file.

Tested on Windows Server 2019.

If the executable is run from the command line directly, it will run normally like before.

Please see it as a first draft based on my multi-year experience with running Go apps as Windows services.

Open questions are:
- Do we need an installer? (WiX, Nullsoft)
- Should the path of the logfile be changeable?
- Should the Windows Event Log be used for logging (requires installer to be feasible)?

Implements https://github.com/GoogleCloudPlatform/cloud-sql-proxy/issues/277

### CODE REVIEW

File: main_windows.go
Comment (enocom) [inline main_windows.go:20]: Let's move all the imports for the proxy and third-party libs below the standard library imports with a blank line separating them.

cf. https://groups.google.com/g/golang-dev/c/fvWvubpTEds

File: windows-service-guide.md
Comment (enocom) [inline windows-service-guide.md:13]: Should we link to the releases page and add a note about installing the 64-bit version?

File: windows-service-guide.md
Comment (enocom) [inline windows-service-guide.md:15]: e.g. `C:\Program Files\cloud-sql-proxy\logs`

File: main_windows.go
Comment (enocom) [inline main_windows.go:82]: This url works for me: https://code.google.com/archive/p/winsvc/issues/4. The one in the comment doesn't for whatever reason.

File: main_windows.go
Comment (enocom) [inline main_windows.go:59]: Should we be setting the arguments on this? How does the command otherwise know to connect to which instances?

**Review by enocom — COMMENTED** [review]
Generally looks good to me.

Tests would be nice ;-)

A few questions and minor suggestions.

**Comment by enocom** [comment]
And to answer you questions:

> Do we need an installer? (WiX, Nullsoft)

For now, installing from the Releases page seems fine to me.

> Should the path of the logfile be changeable?

This is also easy to change in the future, so I'm inclined to start with a hard-coded path and update if there's interest.

> Should the Windows Event Log be used for logging (requires installer to be feasible)?

Might be another nice enhancement, but we can ship this once we work through review. So I'm ok with not doing this yet.

File: windows_install_service.bat
Comment (enocom) [inline windows_install_service.bat:7]: We support `CSQL_PROXY_CREDENTIALS_FILE` for this.

File: windows_install_service.bat
Comment (enocom) [inline windows_install_service.bat:8]: This could just be `CSQL_INSTANCE_CONNECTION_NAME` and doesn't need to be set in the CLI invocation (unless that seems more clear).

See https://github.com/GoogleCloudPlatform/cloud-sql-proxy/blob/main/cmd/root.go#L239.

File: main_windows.go
Comment (enocom) [reply main_windows.go:59]: Maybe we just need a `cmd.SetArgs(args)` here?

**Comment by enocom** [comment]
:laughing: No worries. Again this is a helpful contribution, so I'm happy to lend a hand wherever needed.

**Comment by fkollmann** [comment]
I went through the comments and see no major issues, besides a minor misunderstanding on Windows services. I will answer the comments tomorrow and make some changes.

**Comment by fkollmann** [comment]
Short update: I had to change plans for the weekend. I have time for providing feedback today.

File: main_windows.go
Comment (fkollmann) [reply main_windows.go:20]: np, I literally have no opinion about it. Usually leave it to Goland to handle the imports 🤷

File: main_windows.go
Comment (fkollmann) [reply main_windows.go:59]: I took `cmd.Execute()` as base, and it does not do it, either: https://github.com/GoogleCloudPlatform/cloud-sql-proxy/blob/51d52ba151b09c933ace40874e5af84a63a76665/cmd/root.go#L75

The arguments seem to be taken correctly, by default.

Do you prefer having it changed, anyway?

File: windows_install_service.bat
Comment (fkollmann) [reply windows_install_service.bat:8]: See answer above: https://github.com/GoogleCloudPlatform/cloud-sql-proxy/pull/1696#discussion_r1149711710

**Comment by fkollmann** [comment]
I did some code changes and marked the suggestions as resolved. (Please let me know if I should have kept the suggestions unresolved, or not).

Looking forward to your feedback.

**Comment by enocom** [comment]
Sorry for the slow response. I'll get to this soon.

**Comment by fkollmann** [comment]
Don't worry. No pressure from my side. 

We are currently using it on production, so it at least works for us 😅

**Comment by enocom** [comment]
Currently running this through some manual tests. We can probably get this into the next release.

File: windows-service-guide.md
Comment (enocom) [inline windows-service-guide.md:16]: In my testing, I had to add Network Service to the parent folder as well, but that was because I ran out of my user directory. Should we add a note here about that?

File: windows-service-guide.md
Comment (enocom) [inline windows-service-guide.md:23]: I had to run this in an admin prompt. Should we add that detail?

**Review by enocom — APPROVED** [review]
Looks great. Just a few comments.

Do we need to add a note about UAC as a possible thing to investigate if it doesn't work?

File: windows-service-guide.md
Comment (jackwotherspoon) [inline windows-service-guide.md:3]: nit
```suggestion
This document covers running the *Cloud SQL Auth Proxy* as a service
```

File: windows-service-guide.md
Comment (jackwotherspoon) [inline windows-service-guide.md:60]: potential typo?
```suggestion
Since the Cloud Proxy does not require any file system access, besides the log files,
```

**Comment by enocom** [comment]
@fkollmann There are a few small changes to make here that I'm going to do. And then I'll merge this.

**Comment by enocom** [comment]
Looks like I can't push to your fork. You can change that, though. See [here](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).

Alternatively, aside from the two typos @jackwotherspoon pointed out, we need to keep a `main.go` and just add a `//go:build !windows` at the top, so we can keep building for Unix-based OSes.

**Comment by fkollmann** [comment]
Hi, sry, I was on vacation the last week, incl today. I will look into this tomorrow. Have a nice evening :) . Best Regards, Felix

File: windows_install_service.bat
Comment (fkollmann) [reply windows_install_service.bat:10]: I am not a big fan of adding comments to batch files (mainly because of the bad formatting). However, I added information to the readme file about the settings.

**Comment by fkollmann** [comment]
> Looks like I can't push to your fork. You can change that, though. See [here](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).

That is currently not support with forks below other organizations (instead of individual users). However, I granted you maintainer permissions on the repo.

**Comment by fkollmann** [comment]
Okay, I addressed everything, from my POV. Please feel free to comment (and change). @enocom

**Pull request merged**
